test/common/protobuf/utility_test.cc

```diff
@@ -40,6 +42,65 @@ TEST(UtilityTest, DowncastAndValidate) {
       ProtoValidationException);
 }
 
+TEST(UtilityTest, ValidateDeprecatedFields) {
+  {
+    envoy::test::deprecation_test::Base base;
+    base.set_not_deprecated("foo");
+    // Fatal checks for a non-deprecated field should cause no problem.
+    MessageUtil::checkForDeprecation(base, true);
+  }
+  {
+    envoy::test::deprecation_test::Base base;
+    base.set_is_deprecated("foo");
+    // Non-fatal checks for a deprecated field should log rather than throw an exception.
+    EXPECT_LOG_CONTAINS("warning", "Using deprecated option 'is_deprecated'.",
+                        MessageUtil::checkForDeprecation(base, false));
+    // Fatal checks for a deprecated field should result in an exception.
+    EXPECT_THROW_WITH_REGEX(MessageUtil::checkForDeprecation(base, true), ProtoValidationException,
+                            "Using deprecated option 'is_deprecated'.");
+  }
+  {
+    envoy::test::deprecation_test::Base base;
+    base.mutable_deprecated_message();
+    // Fatal checks for a present (unused) deprecated message should result in an exception.
+    EXPECT_THROW_WITH_REGEX(MessageUtil::checkForDeprecation(base, true), ProtoValidationException,
+                            "Using deprecated option 'deprecated_message'.");
+  }
+
+  {
+    envoy::test::deprecation_test::Base base;
+    base.mutable_not_deprecated_message()->set_inner_not_deprecated("foo");
+    // Non-fatal checks for a deprecated field shouldn't throw an exception.
+    MessageUtil::checkForDeprecation(base, false);
+
+    base.mutable_not_deprecated_message()->set_inner_deprecated("bar");
+    // Fatal checks for a deprecated sub-message should result in an exception.
+    EXPECT_THROW_WITH_REGEX(MessageUtil::checkForDeprecation(base, true), ProtoValidationException,
+                            "Using deprecated option 'inner_deprecated'.");
+  }
+
+  // Check that repeated sub-messages get validated.
```

Comment:
Nit: preference for making these all separate `TEST_F`, since they don't share state and it's more self-documenting.